### DISCUSSION TITLE
Is there community interest in an optional & customizable toolbar?

### DISCUSSION THREAD

**Discussion by kenneth-marut-work** [eclipse-theia/theia#10501] [Show and tell] [answered]
**TLDR:**

*We've built a toolbar for our IDE that's totally optional and customizable. Yes, it strays from VSCode's UI and there are likely varying opinions on this. Is there any interest in the community for us to upstream this feature?*

![tldr](https://user-images.githubusercontent.com/62660714/144643658-61d497c6-7fe8-4f18-95f1-9ac347b0a5bf.PNG)

---

Hi all,

I wanted to gauge interest from the community in upstreaming our optional & customizable toolbar to the Theia respository. This feature has been developed and included in our downstream IDE and we've already seen many users take advantage of it.

I understand that this may be a controversial topic and strays from VSCode's UI/UX but wanted to spark the conversation as there are likely other IDEs that could benefit from this feature.

### **Motivation**

Not all users like keyboard-only access to commands. We've observed that many users of our downstream IDE have found it difficult finding access to basic functionality during the onboarding process (i.e. back+forth) and many users want visual access to frequently repeated tasks (i.e. building).

### **Implementation**

Our toolbar is added to the application shell and can host single icons (executing a command) and custom toolbar widgets which are added through a contribution point. Toolbar layout and customizations are stored in a `toolbar.json` file in `~/.theia`.

### **Features**

**Completely Optional**

The toolbar is enabled or disabled via a preference and can be shown or hidden easily (current keyboard shortcut alt+t). In our IDE, we default it to `off`

![toggle](https://user-images.githubusercontent.com/62660714/144646909-f964f5da-777d-47dc-ab1a-cb6cc8282b0f.gif)

**Default Items and Customizability**

Specify your own defaults to ship with your product. Default icons are customizable (drag and drop to rearrange, remove, add);

*Rearrange/Remove/Group Items:*

![drag-drop-remove](https://user-images.githubusercontent.com/62660714/144643854-98483d79-f4f7-4274-9a44-7f7b1dd62425.gif)

*Add New Items:*

![add-new](https://user-images.githubusercontent.com/62660714/144643873-c6ca0c89-27e0-4ffc-80cf-9860fc9cafca.gif)

**Add Custom Widgets**

Build your own widget using React and bind it to the ReactTabBarToolbarContribution point. Here's an example of a custom Tasks Dropdown that we've built for our IDE

![build-widget](https://user-images.githubusercontent.com/62660714/144647313-ffeaa541-33d5-4950-a75c-66a8956e110d.gif)

---

Feedback and feature requests are welcome!
Thanks 🙂

**Comment by vince-fugnitto** [comment]
I'd be all for it! :+1: The configurable toolbar is a staple in most full-featured IDEs, and if we think about products building applications, they would most likely want the ability to easily consume a toolbar they can configure for their users.

Additional Info:
- https://github.com/eclipse-theia/theia/issues/7540
- https://github.com/eclipse-theia/theia/issues/5037
- https://github.com/microsoft/vscode/issues/41309

**Reply by kenneth-marut-work** [reply]
Great! thanks for the linking those issues 👍

**Comment by ImperialPenguin** [comment]
I'm just getting started and having issues getting Theia to even run. But assuming that gets worked out, I'd love to have access to this toolbar through some means.

**Comment by JonasHelming** [comment]
I simply love this, it would be a great contribution!

**Comment by lucas-koehler** [comment]
Looks awesome. I especially love the idea that you can add any command and select a custom icon. I imagine that really helps users to over time set up an intuitive toolbar that supports their workflow :)

**Comment by ndoschek** [comment]
Looks and sounds awesome! I really like this idea and could immediately think of many use cases where this toolbar would improve the usability of an editor! :+1:

**Comment by ballcoach12** [comment]
+1  for me. We have been discussing how to get this done all morning until we ran across this post.

**Reply by kenneth-marut-work** [reply]
Fantastic, I'll be aiming to get a PR drafted in the following weeks!

**Comment by kenneth-marut-work** [comment, marked as answer]
I have posted a pull request for this feature here #10731.
